ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/StateMachine.hs

```diff
@@ -0,0 +1,744 @@
+
+deriving instance ToExpr TestHash
+deriving instance ToExpr Blk
+deriving instance ToExpr (LedgerState Blk)
+
+{-------------------------------------------------------------------------------
+  Top-level tests
+-------------------------------------------------------------------------------}
+
+type Blk = TestBlock
+type Hdr = TestBlock
+
+prop_sequential :: Property
+prop_sequential = forAllCommands smUnused Nothing $ \cmds -> QC.monadicIO $ do
+    (hist, prop) <- test cmds
+    -- TODO tags
+    prettyCommands smUnused hist prop
+  where
+    test :: QSM.Commands (At Cmd Blk Hdr IO) (At Resp Blk Hdr IO)
+         -> QC.PropertyM IO
+            ( QSM.History (At Cmd Blk Hdr IO) (At Resp Blk Hdr IO)
+            , Property
+            )
+    test cmds = do
+      args <- QC.run $ mkArgs cfg initLedger
+      db   <- QC.run $ openDB args
+      let sm' = sm db genBlock env initLedger
+      (hist, model, res) <- runCommands sm' cmds
+      realChain <- QC.run $ chainHashes <$> toChain db
+      QC.run $ closeDB db
+      let modelChain = chainHashes $ Model.currentChain $ dbModel model
+          prop =
+            counterexample ("Real  chain: " <> show realChain)  $
+            counterexample ("Model chain: " <> show modelChain) $
+            res === Ok .&&. realChain === modelChain
+      return (hist, prop)
+
+    cfg :: NodeConfig (BlockProtocol Blk)
+    cfg = singleNodeTestConfig
+
+    initLedger :: ExtLedgerState Blk
+    initLedger = testInitExtLedger
+
+    env :: ModelEnv Blk Hdr
+    env = ModelEnv { cfg = cfg, blockHeader = id }
+
+    smUnused = sm dbUnused genBlock env initLedger
+
+    -- TODO
+    genBlock :: Gen Blk
+    genBlock = do
+      fork <- choose (1, 4)
+      no   <- choose (1, 20)
+      let prevHash
+            | no == 1   = GenesisHash
+            | otherwise = BlockHash (TestHash (fork * 1000 + no - 1))
+      return TestBlock
+        { tbHash     = TestHash (fork * 1000 + no)
+        , tbPrevHash = prevHash
+        , tbNo       = BlockNo no
+        , tbSlot     = SlotNo no
+        }
+
+    -- Compare the hashes (with '===') so we get much more readable output
+    chainHashes :: Chain Blk -> [Word64]
+    chainHashes = map (unTestHash . tbHash) . chainToList
+
+prop_parallel :: Property
+prop_parallel = forAllParallelCommands smUnused $ \cmds -> QC.monadicIO $
+    test cmds
+  where
+    test :: QSM.ParallelCommands (At Cmd Blk Hdr IO) (At Resp Blk Hdr IO)
+         -> QC.PropertyM IO ()
+    test cmds = do
+      args <- QC.run $ mkArgs cfg initLedger
+      db   <- QC.run $ openDB args
+      let sm' = sm db genBlock env initLedger
+      -- TODO 1 is not enough of course, but otherwise we end up running the
+      -- same commands on the same database.
+      -- IDEA: make sure that the first command is some kind of openDB call
+      -- and the last command is some kind of closeDB call.
+      prettyParallelCommands cmds =<< runParallelCommandsNTimes 1 sm' cmds
```

Comment:
Make sure that the first command is just a call to `Reset`, which closes, resets the FS, and opens the DB. We can just prepend that command here, so the shrinker won't get its hands on it.